src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,434 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+)
+
+// dcAndPrivateKey stores a delegated credential and its corresponding private
+// key.
+type dcAndPrivateKey struct {
+	*DelegatedCredential
+	privateKey crypto.PrivateKey
+}
+
+// These test keys were generated with the following program, available in the
+// crypto/tls directory:
+//
+//		go run generate_cert.go -ecdsa-curve P256 -host 127.0.0.1 -dc
+//
+var delegatorCertPEM = `-----BEGIN CERTIFICATE-----
+MIIBdzCCAR2gAwIBAgIQSq4WxfX6RiAGB184js+2mTAKBggqhkjOPQQDAjASMRAw
+DgYDVQQKEwdBY21lIENvMB4XDTIwMDgzMDA2MjU1OVoXDTIxMDgzMDA2MjU1OVow
+EjEQMA4GA1UEChMHQWNtZSBDbzBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABK4S
+bedwIrqjB+WYQgc9b1gcj4Cpx9Qk8DgspkjnWNXSuQXogVdimGKawUnsha5r0bEk
+/faBwFfLXd8sPJ+8i6+jVTBTMA4GA1UdDwEB/wQEAwIHgDATBgNVHSUEDDAKBggr
+BgEFBQcDATAMBgNVHRMBAf8EAjAAMA0GCSsGAQQBgtpLLAQAMA8GA1UdEQQIMAaH
+BH8AAAEwCgYIKoZIzj0EAwIDSAAwRQIhAJ6d74ipcNAXk6FO/GWaILxg+Ng4jIbN
+9/XX2GvEwUyAAiAOlG4G7uHKhCmBnqpHlymndz5hcr6Ce/5DFQJsMDEc2g==
+-----END CERTIFICATE-----
+`
+
+var delegatorKeyPEM = `-----BEGIN EC PRIVATE KEY-----
+MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgwUuTsex2vppn2HYY
+0kYJu7MbWc72IJgab1uNxkJN+JuhRANCAASuEm3ncCK6owflmEIHPW9YHI+AqcfU
+JPA4LKZI51jV0rkF6IFXYphimsFJ7IWua9GxJP32gcBXy13fLDyfvIuv
+-----END EC PRIVATE KEY-----
+`
+
+var nonDelegatorCertPEM = `-----BEGIN CERTIFICATE-----
+MIIBaDCCAQ6gAwIBAgIQcMnAGu3NQYTGYf2HK+JodTAKBggqhkjOPQQDAjASMRAw
+DgYDVQQKEwdBY21lIENvMB4XDTIwMDgxODA1NDg1NloXDTIxMDgxODA1NDg1Nlow
+EjEQMA4GA1UEChMHQWNtZSBDbzBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABPAi
+QzOthHUdwLTPo9P7Vk1I2W5RHW5nIkq9zYqqMZ5mHQ6vmmrpklvTNHtY93PlokjN
+pnlhzEsxK/QrBoAQ8fajRjBEMA4GA1UdDwEB/wQEAwIFoDATBgNVHSUEDDAKBggr
+BgEFBQcDATAMBgNVHRMBAf8EAjAAMA8GA1UdEQQIMAaHBH8AAAEwCgYIKoZIzj0E
+AwIDSAAwRQIgbOxx7/KWTD47UTWIBcFB95BPrFp2SaFBUyjhzMDXsQkCIQDnwtye
+V1OlcMigjCsQuGRacYFP3f1ASpYVv58t/ZeVCw==
+-----END CERTIFICATE-----
+`
+
+var nonDelegatorKeyPEM = `-----BEGIN EC PRIVATE KEY-----
+MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgD9Q9131NamLDe4ud
+dU9rg+gO0vv8lXYErf7P5GQlZD6hRANCAATwIkMzrYR1HcC0z6PT+1ZNSNluUR1u
+ZyJKvc2KqjGeZh0Or5pq6ZJb0zR7WPdz5aJIzaZ5YcxLMSv0KwaAEPH2
+-----END EC PRIVATE KEY-----
+`
+
+var dcTestConfig *Config
+var dcTestCerts map[string]*Certificate
+var dcTest *dcAndPrivateKey
+var dcNow time.Time
+var dcTestDCScheme = ECDSAWithP521AndSHA512
```

Comment:
It would be great if this would be parameterized for the tests, so we can test each DC scheme.

Reply:
Good idea.

Reply:
Added now. ;)